compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt

```diff
@@ -0,0 +1,405 @@
+            if (hasXPart) {
+                CupertinoOverscrollDirection.HORIZONTAL
+            } else {
+                // hasYPart != hasXPart and hasXPart is false
+                CupertinoOverscrollDirection.VERTICAL
+            }
+        } else {
+            // hasXPart and hasYPart are equal
+            CupertinoOverscrollDirection.UNKNOWN
+        }
+    }
+
+    private fun CupertinoOverscrollDirection.combinedWith(other: CupertinoOverscrollDirection): CupertinoOverscrollDirection =
+        when (this) {
+            CupertinoOverscrollDirection.UNKNOWN -> when (other) {
+                CupertinoOverscrollDirection.UNKNOWN -> CupertinoOverscrollDirection.UNKNOWN
+                CupertinoOverscrollDirection.VERTICAL -> CupertinoOverscrollDirection.VERTICAL
+                CupertinoOverscrollDirection.HORIZONTAL -> CupertinoOverscrollDirection.HORIZONTAL
+            }
+
+            CupertinoOverscrollDirection.VERTICAL -> when (other) {
+                CupertinoOverscrollDirection.UNKNOWN, CupertinoOverscrollDirection.VERTICAL -> CupertinoOverscrollDirection.VERTICAL
+                CupertinoOverscrollDirection.HORIZONTAL -> CupertinoOverscrollDirection.HORIZONTAL
+            }
+
+            CupertinoOverscrollDirection.HORIZONTAL -> when (other) {
+                CupertinoOverscrollDirection.UNKNOWN, CupertinoOverscrollDirection.HORIZONTAL -> CupertinoOverscrollDirection.HORIZONTAL
+                CupertinoOverscrollDirection.VERTICAL -> CupertinoOverscrollDirection.VERTICAL
+            }
+        }
+    private fun Velocity.toOffset(): Offset =
+        Offset(x, y)
+
+    private fun Offset.toVelocity(): Velocity =
+        Velocity(x, y)
+
+    private fun Velocity.toFloat(): Float =
+        toOffset().toFloat()
+
+    private fun Float.toVelocity(): Velocity =
+        toOffset().toVelocity()
+
+    private fun Offset.toFloat(): Float =
+        when (direction) {
+            CupertinoOverscrollDirection.UNKNOWN -> 0f
+            CupertinoOverscrollDirection.VERTICAL -> y
+            CupertinoOverscrollDirection.HORIZONTAL -> x
+        }
+
+    private fun Float.toOffset(): Offset =
+        when (direction) {
+            CupertinoOverscrollDirection.UNKNOWN -> Offset.Zero
+            CupertinoOverscrollDirection.VERTICAL -> Offset(0f, this)
+            CupertinoOverscrollDirection.HORIZONTAL -> Offset(this, 0f)
+        }
+
+    private suspend fun playInitialSpringAnimationIfNeeded(initialVelocity: Velocity): Velocity {
+        val velocity = initialVelocity.toFloat()
+        val overscroll = overscrollOffset.toFloat()
+
+        return if ((velocity < 0f && overscroll > 0f) || (velocity > 0f && overscroll < 0f)) {
+            playSpringAnimation(0f, velocity, true).toVelocity()
+        } else {
+            initialVelocity
+        }
+    }
+
+    private suspend fun playSpringAnimation(
+        unconsumedDelta: Float,
+        initialVelocity: Float,
+        flingFromOverscroll: Boolean
```

Comment:
There is an opinion that such flag parameters are bad. It's more discussion topic, I'm not going to block anything because of it

> Robert C. Martin's Clean Code Tip 12: Eliminate Boolean Arguments
> Boolean arguments loudly declare that the function does more than one thing. They are confusing and should be eliminated.

Reply:
Agree, refactored